Restore caching cleanup

dannyroberts opened the pull request:
To fix https://manage.dimagi.com/default.asp?260109

This PR essentially picks up https://github.com/dimagi/commcare-hq/pull/17523 (which got reverted). I start by reverting the revert, which had a number of conflicts with some of @millerdev's recent PRs, which I fixed.

@millerdev I've addressed your comments here https://github.com/dimagi/commcare-hq/pull/17523#pullrequestreview-57380590. Notably, I took your changes to `test_sync_mode.py` over mine and then fixed a couple things in subsequent commits (particularly https://github.com/dimagi/commcare-hq/pull/17554/commits/d5c952d3f5e1856c9b256615fad2418c40c6b069).

In addition to the tests running on travis, I have been testing this using a modified version of a test that clayton wrote for this issue, which is on a not-to-be-merged branch of commcare-core, https://github.com/dimagi/commcare-core/tree/danny-local/async_collison_replication. (You can see what I had to change to get it to run against my local dev setup here https://github.com/dimagi/commcare-core/compare/ctsims/async_collison_replication...danny-local/async_collison_replication?w=1).

The commits that are actually responsible for fixing clayton's test are
- https://github.com/dimagi/commcare-hq/pull/17554/commits/99074d1153fba7b937e65bf3017c29228771488e
- https://github.com/dimagi/commcare-hq/pull/17554/commits/7601a9cb527b013cc19b14623ba33d8f6413dd63
- https://github.com/dimagi/commcare-hq/pull/17554/commits/a7897e965331278db2da869ad3209f220fbf0c08

dannyroberts committed:
move restore caching helpers to new file restore_caching.py

dannyroberts committed:
replace async_restore_task_id_cache_key with AsyncRestoreTaskIdCache

cache accessor class that lightly abstracts getting and setting the cache value

dannyroberts committed:
replace restore_payload_path_cache_key with RestorePayloadPathCache

cache accessor class that lightly abstracts getting and setting the cache value

dannyroberts committed:
remove get_redis_default_cache aliases

dannyroberts committed:
update mocks to fix tests

dannyroberts committed:
more aggressively invalidate initial restore

dannyroberts committed:
invalidate restore caches with and without device id on submit

dannyroberts committed:
add debugging to restore cache accessors

dannyroberts committed:
invalidate all restore cache keys

dannyroberts committed:
Revert "Merge pull request #17530 from dimagi/revert-17523-fix-restore-payload-cache-test"

This reverts commit 92cb42845be349c32059cc0d8a1d316b82e60bbc, reversing
changes made to ab87de05b6e5d88f17eb784ff4f928d19d30bc69.

dannyroberts committed:
don't test different sync versions have different caches

because they no longer do, because V1 syncs aren't super supported,
and this appeared to be somewhat broken anyway

dannyroberts committed:
fix tests after complex revert

dannyroberts committed:
Merge branch 'master' into restore-caching-cleanup

dannyroberts committed:
lint from stickler review

dannyroberts committed:
remove no-longer-relevant comments

dannyroberts committed:
use domain instead of response_class in cache key

ctsims commented:
Thanks, Danny! @millerdev This definitely could use a review ASAP, this is the data collision fix I was mentioning.

dannyroberts commented:
(cc: @nickpell new buddy)

millerdev commented on corehq/ex-submodules/casexml/apps/phone/restore_caching.py:
```diff
@@ -0,0 +1,75 @@
+class RestorePayloadPathCache(CacheAccessor):
+    def __init__(self, domain, user_id, sync_log_id, device_id):
+        self.cache_key = _restore_payload_path_cache_key(domain, user_id, sync_log_id, device_id)
+        self.debug_info = ('RestorePayloadPathCache', domain, user_id, sync_log_id, device_id)
```
Do the values in this cache never expire? Is it possible that some key/value pairs would never be deleted and could that become a problem?

dannyroberts replied:
do you ask because there's no timeout set? My intention is for that to meant that it defaults back to the django cache's default, which is what it was doing before (at timeout wasn't being passed in when restore payload path was being set in the cache). I thought we had a default timeout in that case, but you might be right that that just means never expire. Anyway, seems reasonable to only cache it for a day like the other one, so I'll make that change.

millerdev commented on corehq/ex-submodules/casexml/apps/phone/restore_caching.py:
```diff
@@ -0,0 +1,75 @@
+        domain=domain,
+        prefix=RESTORE_CACHE_KEY_PREFIX,
+        user_id=user_id,
+        version='2.0',
```
what does this version mean? why is it hard-coded here rather than using the version specified in the restore config?

dannyroberts replied:
It's just left over and never passed in, I'll just remove it

millerdev reviewed (approved):
Few comments, probably not blocking.

dannyroberts commented on corehq/ex-submodules/casexml/apps/phone/restore_caching.py:
```diff
@@ -0,0 +1,75 @@
+
+def _restore_cache_key(domain, prefix, user_id, version, sync_log_id, device_id):
+    # to invalidate all restore cache keys, increment the number below
+    hashable_key = '8-{domain}-{prefix}-{user}-{version}-{sync_log_id}-{device_id}'.format(
```
Yeah, good point. Honestly I picked 8 because every time I wanted to try it fresh locally i would bump it, and I ended up with 8 there.

I'll change the dashes to commas.

dannyroberts committed:
clean up _restore_cache_key a bit

including using a separator character that isn't likely to appear in any
of the fields, so as to avoid any collisions however unlikely

dannyroberts committed:
give RestorePayloadPathCache a timeout of 1 day

millerdev commented on corehq/ex-submodules/casexml/apps/phone/restore_caching.py:
```diff
@@ -24,7 +23,6 @@ def _restore_payload_path_cache_key(domain, user_id, sync_log_id, device_id):
         domain=domain,
         prefix=RESTORE_CACHE_KEY_PREFIX,
         user_id=user_id,
-        version='2.0',
```
Hmm, that's not quite what I had in mind. It's not clear to me why we don't want the sync protocol version in the cache key. Can you help me understand your reasoning here?

dannyroberts replied:
Yeah, so the reasoning is, it looked to me like it wasn't being done consistently anyway, so having it look like it was working isn't as good as just not having it there at all; as to the option to fix it, V1 is very, very deprecated, so there's really only one version, 2.0. I'm not too worried about the use case of someone creating a new app on their old domain from 2012 and installing that on the phone of one of their users who recently restored.

dannyroberts replied:
(It may be time to get rid of v1 sync entirely, actually)

millerdev replied:
I'm very pro getting rid of V1. However, I thought I saw [references](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore_v3.py) to [V3](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/case/xml/__init__.py#L7) in the code. Is that not a thing or was that a false start?

dannyroberts replied:
Yeah I saw that. I think it's a false start, but @ctsims can confirm

ctsims replied:
V3 is highly likely to occur at some point, but there are definitely no short term plans to migrate to a new sync format.
